From 852b162c8255b2bdecde76e89e272149548e0669 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 7 Nov 2018 09:31:17 +0000
Subject: [PATCH] Allocate space in structs pre-emptively to increase chances
 of livepatching

1. Expand FEATURESET for 4 more words
2. Increment CPUID_GUEST_NR_FEAT to get an additional feature leaf
   in struct cpuid_policy
3. Reserve space in the following structs:

    grant_table
    intel_iommu
    iommu
    amd_iommu
    guest_iommu
    hvm_ioreq_vcpu
    hvm_ioreq_server
    msr_domain_policy
    msr_vcpu_policy
    p2m_domain

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index 281ca2e..64a905c 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -80,6 +80,9 @@ struct grant_table {
     const struct domain *domain;
 
     struct grant_table_arch arch;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 unsigned int __read_mostly opt_max_grant_frames = 64;
diff --git a/xen/drivers/passthrough/vtd/iommu.h b/xen/drivers/passthrough/vtd/iommu.h
index 1a992f7..d34cb14 100644
--- a/xen/drivers/passthrough/vtd/iommu.h
+++ b/xen/drivers/passthrough/vtd/iommu.h
@@ -531,6 +531,9 @@ struct intel_iommu {
     struct ir_ctrl ir_ctrl;
     struct iommu_flush flush;
     struct acpi_drhd_unit *drhd;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 struct iommu {
@@ -548,6 +551,9 @@ struct iommu {
     struct list_head ats_devices;
     unsigned long *domid_bitmap;  /* domain id bitmap */
     u16 *domid_map;               /* domain id mapping array */
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 static inline struct qi_ctrl *iommu_qi_ctrl(struct iommu *iommu)
diff --git a/xen/include/asm-x86/amd-iommu.h b/xen/include/asm-x86/amd-iommu.h
index 02715b4..c4b5d70 100644
--- a/xen/include/asm-x86/amd-iommu.h
+++ b/xen/include/asm-x86/amd-iommu.h
@@ -106,6 +106,9 @@ struct amd_iommu {
     int enabled;
 
     struct list_head ats_devices;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 struct ivrs_mappings {
@@ -183,6 +186,9 @@ struct guest_iommu {
 
     /* guest interrupt settings */
     struct guest_iommu_msi  msi;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 extern bool_t iommuv2_enabled;
diff --git a/xen/include/asm-x86/hvm/domain.h b/xen/include/asm-x86/hvm/domain.h
index 3e73318..5e035d3 100644
--- a/xen/include/asm-x86/hvm/domain.h
+++ b/xen/include/asm-x86/hvm/domain.h
@@ -46,6 +46,9 @@ struct hvm_ioreq_vcpu {
     struct vcpu      *vcpu;
     evtchn_port_t    ioreq_evtchn;
     bool             pending;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[1];
 };
 
 #define NR_IO_RANGE_TYPES (XEN_DMOP_IO_RANGE_PCI + 1)
@@ -67,6 +70,9 @@ struct hvm_ioreq_server {
     struct rangeset        *range[NR_IO_RANGE_TYPES];
     bool                   enabled;
     uint8_t                bufioreq_handling;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 /*
diff --git a/xen/include/asm-x86/msr.h b/xen/include/asm-x86/msr.h
index 51d1bed..34acfd6 100644
--- a/xen/include/asm-x86/msr.h
+++ b/xen/include/asm-x86/msr.h
@@ -305,6 +305,9 @@ struct vcpu_msrs
      * values here may be stale in current context.
      */
     uint32_t dr_mask[4];
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 void init_guest_msr_policy(void);
diff --git a/xen/include/asm-x86/p2m.h b/xen/include/asm-x86/p2m.h
index 3304921..05f51e9 100644
--- a/xen/include/asm-x86/p2m.h
+++ b/xen/include/asm-x86/p2m.h
@@ -362,6 +362,9 @@ struct p2m_domain {
          unsigned int flags;
          unsigned long entry_count;
      } ioreq;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 /* get host p2m table */
diff --git a/xen/include/xen/lib/x86/cpuid.h b/xen/include/xen/lib/x86/cpuid.h
index 22d43ef..7fcc63b 100644
--- a/xen/include/xen/lib/x86/cpuid.h
+++ b/xen/include/xen/lib/x86/cpuid.h
@@ -66,7 +66,7 @@ static inline void cpuid_count_leaf(
 #undef XCHG
 
 #define CPUID_GUEST_NR_BASIC      (0xdu + 1)
-#define CPUID_GUEST_NR_FEAT       (0u + 1)
+#define CPUID_GUEST_NR_FEAT       (1u + 1)
 #define CPUID_GUEST_NR_CACHE      (5u + 1)
 #define CPUID_GUEST_NR_TOPO       (1u + 1)
 #define CPUID_GUEST_NR_XSTATE     (62u + 1)
@@ -258,6 +258,8 @@ struct cpuid_policy
 static inline void cpuid_policy_to_featureset(
     const struct cpuid_policy *p, uint32_t fs[FEATURESET_NR_ENTRIES])
 {
+    unsigned int i;
+
     fs[FEATURESET_1d]  = p->basic._1d;
     fs[FEATURESET_1c]  = p->basic._1c;
     fs[FEATURESET_e1d] = p->extd.e1d;
@@ -268,6 +270,10 @@ static inline void cpuid_policy_to_featureset(
     fs[FEATURESET_e7d] = p->extd.e7d;
     fs[FEATURESET_e8b] = p->extd.e8b;
     fs[FEATURESET_7d0] = p->feat._7d0;
+
+    /* Zero reserved featureset words */
+    for ( i = FEATURESET_7d0 + 1; i < FEATURESET_NR_ENTRIES; i++ )
+        fs[i] = 0;
 }
 
 /* Fill in a CPUID policy from a featureset bitmap. */
diff --git a/xen/include/xen/lib/x86/msr.h b/xen/include/xen/lib/x86/msr.h
index e2cfbb1..d5a1290 100644
--- a/xen/include/xen/lib/x86/msr.h
+++ b/xen/include/xen/lib/x86/msr.h
@@ -23,6 +23,9 @@ struct msr_policy
             bool cpuid_faulting:1;
         };
     } plaform_info;
+
+    /* Reserved space */
+    uint64_t _rsvd_livepatch[2];
 };
 
 #ifdef __XEN__
diff --git a/xen/tools/gen-cpuid.py b/xen/tools/gen-cpuid.py
index 1eea4c3..afeca52 100755
--- a/xen/tools/gen-cpuid.py
+++ b/xen/tools/gen-cpuid.py
@@ -130,6 +130,8 @@ def crunch_numbers(state):
 
     # Size of bitmaps
     state.nr_entries = nr_entries = (max(state.names.keys()) >> 5) + 1
+    reserved_entries_num = 4
+    state.nr_entries = nr_entries = nr_entries + reserved_entries_num
 
     # Features common between 1d and e1d.
     common_1d = (FPU, VME, DE, PSE, TSC, MSR, PAE, MCE, CX8, APIC,
